core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalCheckpointThread.java

```diff
@@ -224,6 +246,10 @@ private void runInternal() {
 
       // Sleep for a while if no entry is found.
       if (entry == null) {
+        if (continuousSleepNum >= mReplayDoneSleepThreshold) {
```

Comment:
Why do we need to wait for multiple sleeps?

Reply:
I cannot eliminate the chance that the alluxio cluster has some distance from its journal system (HDFS) and maybe at some period of time the journal transfer is slow. 
Like send ... send ... send ... send ..  (1 sec gap)... send 
waiting for a longer sleep time can better make us confident that the journal replay is actually done.

Reply:
Wouldn't `UfsJournalCheckpointThread` only sleep when it already read the last entry that have been written recently? Is that not enough indicator that it has caught up all the entries that has been written by the primary master at the time of startup?

Reply:
Oh you should be correct, i don't need to sleep for that long time, one sleep is enough,

Reply:
@bf8086 I tested if i directly set replay == DONE when entry == null, the journal replay will be considered as finished before the master reading any journals. 

The logs look like
```
2021-02-02 03:32:43,590 INFO  ZkMasterInquireClient - All masters: [ReplayUfsJournalZK4-masters-2:19998]
2021-02-02 03:34:05,495 INFO  JournalUtils - Reading journal entries
2021-02-02 03:34:05,497 INFO  JournalUtils - Reading journal entries
2021-02-02 03:34:05,497 INFO  JournalUtils - Reading journal entries
2021-02-02 03:34:05,497 INFO  JournalUtils - Reading journal entries
2021-02-02 03:34:06,277 INFO  UfsJournalSystem - Finish master process ufs journal initial replay in 90528 ms
2021-02-02 03:34:06,277 INFO  FaultTolerantAlluxioMasterProcess - Journal initial replay finished
2021-02-02 03:34:07,211 INFO  UfsJournalReader - Reading journal file hdfs://ReplayUfsJournalZK4-masters-1:9000/alluxio_journal/FileSystemMaster/v1/logs/0x858b9b-0x861492.
2021-02-02 03:34:07,483 INFO  UfsJournalReader - Reading journal file hdfs://ReplayUfsJournalZK4-masters-1:9000/alluxio_journal/FileSystemMaster/v1/logs/0x861492-0x869c2f.
2021-02-02 03:34:07,939 INFO  UfsJournalReader - Reading journal file hdfs://ReplayUfsJournalZK4-masters-1:9000/alluxio_journal/FileSystemMaster/v1/logs/0x869c2f-0x872793.
```

Reply:
The UfsJournalCheckpointThread could sleep if it hasn't read the HDFS journal fast enough.

Reply:
The logs look like it got entry=null when it finished reading the checkpoint. So a more accurate indicator for journal read completion is the return value of `mJournalReader.advance()`. It would be something like:
```
      try {
        switch (mJournalReader.advance()) {
          case CHECKPOINT:
            ...
          case ENTRY:
            ...
          case DONE:
            mReplayState = ReplayState.REPLAY_DONE;
            break;
          default:
            break;
        }
```